### PR TITLE
Bugfix/amaroc 681 double layers

### LINKED ISSUE #107: [Bug]: grouping step creates overlapping groups
### Is there already an issue for this?

- [X] I found no existing issue for this bug.

### Describe the bug

The grouping step currently sets no restrictions on the vertical separation between the groups. In certain circumstances, groups can be found to have the exact same altitude, as for example in Geneva on 2018-04-30 @ 18:17 UTC (see example below). 

This behavior is a direct and unwanted behavior of the current grouping algorithm. Each sub-group identified is assigned to the slice where most of its points belong. In the example below, the majority of the points are found to be linked (according to the single linkage criterion) and assigned to the pink slice where the majority of them belong. A handful of isolated blue points are then all put inside one group assigned to the blue slice. Very clearly, however, the two groups are overlapping with each other.  

**Solution (?):**

A good option could be to verify that the groups identified are separated by some minimal distance (possibly a function of altitude) as is done for the layering step. The layering "validation" approach (see l.252-265 in layer.py) could be used for the grouping step - either by duplicating it in the grouping routine, or by extracting it to a higher level. Alternatively (and perhaps better, since it would not require any new parameter), **one could simply request that groups cannot be overlapping in terms of their vertical extent** - and if they are, to merge them.

There is the idea of a final verification step that all cloud layers are sufficiently offset from one another (see #88), which would help mask this bug. I suspect however that it would be best to fix the grouping step itself, rather than mask its deficiency with a subsequent correction. **The idea that groups cannot overlap one another would be easy to implement and quite meaningful - I would probably try this first ?**

### Steps to reproduce

Run the normal algorithm for the case above.

### Relevant log output

```shell
n.a.
```

### Environment

```markdown
ampycloud: v0.6.0.dev0
```

### Code of Conduct

- [X] I agree to follow the ampycloud Code of Conduct.

### PR DESCRIPTION
**Pull request by regDaniel**
**Description:**

This PR does two things in order to resolve #107
1. Introduce minimum separation to grouping step.
2. Change the layer separation criterion from mean altitude to base altitude.

Some refactoring was necessary to achieve the above
[] Split up metarize method to separate creation of df, layer height calculation, layer amount calculation and creation of METAR codes.
[] make the two minimum separation parameters global parametes instead of layering parameters.
[] move the base height calculation to a utility function.

**Error(s) fixed:**
#107 

**Checklists**:
- [ x] New code includes dedicated tests.
- [x] New code has been linted.
- [x] New code follows the project's style.
- [x] New code is compatible with the 3-Clause BSD license.
- [x] CHANGELOG has been updated.
- [x] AUTHORS has been updated.
- [x] Copyright years in module docstrings have been updated.

### CODE REVIEW

File: src/ampycloud/layer.py [hunk 1]
```diff
@@ -240,8 +241,27 @@ def ncomp_from_gmm(vals: np.ndarray,
...
+        n_largest_idxs = vals.argsort()[
+            - int(len(vals) * lookback_perc / 100):
+        ]
+        n_largest_elements = vals[n_largest_idxs]
+        if len(n_largest_elements) == 0:
+            warnings.warn(
+                'Cloud base calculation got an empty array.'
+                'Maybe check lookback percentage (is set to %i)' %lookback_perc
+            )
+            return np.nan
```
Comment (regDaniel) [inline src/ampycloud/layer.py:255]: @fpavogt the quantile function raised when I passed it an empty array. The `np.mean` we used before would return a nan and I assume this is what happened before my change. But two things are unclear to me:
1.Is it acceptable to return a nan as cloud base height for a given layer? Do we treat that case somewhere (metarize maybe)?
2. How can we have an empty array here? Seems 🐟 -y no?

Comment (regDaniel) [reply src/ampycloud/layer.py:255]: I should add, the test that failed was `test_identical_pts()` in `test_layers.py` and therein when we have two distinct but close points.

Comment (fpavogt) [reply src/ampycloud/layer.py:255]: Agreed, I would not expect an empty array to be possible. I'll checkout your branch and take a close look. One option is that "non-detection hits" (i.e. hits with type 0 and a NaN altitude) are included somehow ? Just a thought ?

Comment (regDaniel) [reply src/ampycloud/layer.py:255]: I will quickly check if my branch introduced the empty array or if it was there before but went through the `np.mean` unnoticed. That could allow us to pinpoint the issue better.

Comment (fpavogt) [reply src/ampycloud/layer.py:255]: Ok, here is what happens. The empty layer arises in this test because we feed a list of hits with only two distinct values, but search for 3 Gaussian components, which somehow ends up being the preferred amount of components returned by the GMM. 

In operations, this would never happen because after #78, a safety check was included in data.py (inside find_layers() on l.755+). 

My proposition would therefore be the following:
- on the new _calc_base_height(), keep your check for len(n_largest_element) == 0, but raise an error instead of a warning if it happens, because a NaN makes no sense.
- in ncomp_from_gmm(), around l.194, I would add a check for the number of individual values, and limit ncomp_max to this (directly after the check if there's only 1 value):

```
    elif len(np.unique(vals_orig)) < ncomp_max:
        ncomp_max = len(np.unique(vals_orig))
        logger.warning(f'Restricting ncomp_max to the max number of individual values: {ncomp_max}')
```

- I would not change data.py and keep the existing ncomp test in there

File: test/ampycloud/ref_data/debug107_2020.01.01-00.01.01_SCT024-SCT031.csv
Comment (fpavogt) [inline test/ampycloud/ref_data/debug107_2020.01.01-00.01.01_SCT024-SCT031.csv:1]: Is this example used in any of the tests ? The naming convention of the files implies that it is not checked for scientific validity. If it is not used, might be best not to include it in the PR. But if it is deemed useful, it would be good to add either a dedicated test, or to change the filename so that it gets picked up by `test_scientific_validty.py`

File: test/ampycloud/ref_data/debug107_2020.01.01-00.16.01_BKN010-BKN025.csv
Comment (fpavogt) [inline test/ampycloud/ref_data/debug107_2020.01.01-00.16.01_BKN010-BKN025.csv:1]: Same comment as above for the other `debug107_...` mock data file.

File: test/ampycloud/ref_data/debug107_2020.01.01-00.01.01_SCT024-SCT031.csv
Comment (fpavogt) [reply test/ampycloud/ref_data/debug107_2020.01.01-00.01.01_SCT024-SCT031.csv:1]: Nevermind this comment, the filename already works just fine.

**Comment by regDaniel** [comment]
@fpavogt The PR is now almost ready for review and if you want you can start with it whenever you feel like. The only thing I still plan to add, is a unit test for the group splitting on Monday (now it's only tested via the scientific tests). I went for a "minimum effort" refactoring of metarize, given our planned deployment to DEPL next week. Basically, the metarize method is chunked into components that seemed logically to me. These are then wrapped in private methods and called inside metarize. On top of that I can use these methods for the group separation. It's not the most elegant solution but it avoids most unnecessary calculations.

File: test/ampycloud/ref_data/Geneva_2018.04.30-18.17.00_SCT100-BKN120.csv [hunk 2]
```diff
@@ -0,0 +1,274 @@
+,ceilo,dt,alt,type
```
Comment (fpavogt) [inline test/ampycloud/ref_data/Geneva_2018.04.30-18.17.00_SCT100-BKN120.csv:1]: The first column with the indexes is not required. Removing it will remove a warning when running the tests.

File: test/ampycloud/ref_data/Geneva_2020.10.22.-21.15.00_SCT004-SCT007.csv [hunk 3]
```diff
@@ -0,0 +1,249 @@
+,ceilo,dt,alt,type
+0,Ceilometer.Final05,-886.0,570.0,1
```
Comment (fpavogt) [inline test/ampycloud/ref_data/Geneva_2020.10.22.-21.15.00_SCT004-SCT007.csv:2]: Same as previously. Would be good to remove the indexes.

**Review by fpavogt — APPROVED** [review]
Globally looks fine by me. Just a handful of suggestions here and there, but nothing critical.

**Review by MericRed — COMMENTED** [review]
Sometimes there are named parameters without a type specification. Perhaps these could be added (where helpful), along with the pydoc that @fpavogt mentioned?

**Review by MericRed — APPROVED** [review]

File: test/ampycloud/ref_data/Geneva_2018.04.30-18.17.00_SCT100-BKN120.csv (hunk 2 above)
Comment (regDaniel) [reply test/ampycloud/ref_data/Geneva_2018.04.30-18.17.00_SCT100-BKN120.csv]: Done, thanks!

File: test/ampycloud/ref_data/Geneva_2020.10.22.-21.15.00_SCT004-SCT007.csv (hunk 3 above)
Comment (regDaniel) [reply test/ampycloud/ref_data/Geneva_2020.10.22.-21.15.00_SCT004-SCT007.csv]: Done, thanks!

File: src/ampycloud/layer.py [hunk 4]
```diff
@@ -240,8 +244,27 @@ def ncomp_from_gmm(vals: np.ndarray,
...
+        n_largest_idxs = vals.argsort()[
+            - int(len(vals) * lookback_perc / 100):
+        ]
+        n_largest_elements = vals[n_largest_idxs]
+        if len(n_largest_elements) == 0:
+            raise ValueError(
+                'Cloud base calculation got an empty array.'
+                'Maybe check lookback percentage (is set to %i)' %lookback_perc
+            )
+            return np.nan
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply src/ampycloud/layer.py]: Absolutely, changed it, thanks!

File: src/ampycloud/data.py [hunk 5]
```diff
@@ -395,35 +372,33 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
...
-            # Start computing the base altitude
-            # First, compute which points should be considered in terms of lookback time
-            n_to_use = int(np.floor(len(self.data.loc[in_sligrolay]) * base_lvl_lookback_perc/100))
-            # Then, actually compute the base altitude, possibly ignoring the lowest points
-            pdf.iloc[ind, pdf.columns.get_loc('alt_base')] = \
-                np.percentile(self.data.loc[in_sligrolay].nlargest(n_to_use, 'dt')['alt'],
-                              base_lvl_alt_perc)
+        return pdf
 
+    def _add_sligrolay_information(self, which, pdf, original_ids):
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply src/ampycloud/data.py]: I added docstrings to the private methods. It's probably a good idea 🙂

**Comment by regDaniel** [comment]
> Sometimes there are named parameters without a type specification. Perhaps these could be added (where helpful), along with the pydoc that @fpavogt mentioned?

Thanks for catching it! I added the type hints, where they were still missing. I wonder if we could configure a linter to catch that automatically. If you know @fpavogt let me know, otherwise I'll check tomorrow.

**Comment by fpavogt** [comment]
> I wonder if we could configure a linter to catch that automatically. If you know @fpavogt let me know, otherwise I'll check tomorrow.

Sorry @regDaniel, I've got no clue on this.

**Comment by regDaniel** [comment]
> > I wonder if we could configure a linter to catch that automatically. If you know @fpavogt let me know, otherwise I'll check tomorrow.
> 
> Sorry @regDaniel, I've got no clue on this.

Ok I found two things:

1. We can enforce using type hints by passing `--disallow-untyped-defs` to mypy.
2. We do not use mypy in the actions workflow. Sorry, I was implicitly assuming we use it already and I would strongly encourage us to do so. If I run it locally I get quite some static type checking errors. Those are not necessarily bugs but we need to fix this, I'll address the most severe ones now and open an issue for the rest. @loforest if we fix typing errors at a later stage, can we deliver a patch release without CN?

File: src/ampycloud/data.py [hunk 6]
```diff
@@ -354,36 +391,47 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
...
+                    raise AmpycloudError(
+                        'Layering already done.'
+                        ' If you metarize your groups now, you will loose the'
+                        ' layering information !'
+                    )
                 # If all is as expected, then set the number of sub-components to -1 for now, until
                 # the layering step decides otherwise (possibly).
                 pdf.loc[ind, 'ncomp'] = -1
+            # Keep track of the original sli/gro/lay id
+            pdf.loc[ind, 'original_id'] = oid
```
Comment (loforest) [inline src/ampycloud/data.py:411]: In the previous version, this line was before the `if which == 'groups':` (line 397). Is this ok?

File: src/ampycloud/data.py [hunk 7]
```diff
@@ -354,36 +391,47 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
...
+
+        return pdf, original_ids
+
+    def _calculate_cloud_cover(
```
Comment (loforest) [inline src/ampycloud/data.py:415]:
```suggestion
    def _calculate_cloud_amount(
```

File: src/ampycloud/data.py [hunk 8]
```diff
@@ -354,36 +391,47 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
...
+            self, which: str, pdf: pd.DataFrame, original_ids: npt.ArrayLike
+        ) -> pd.DataFrame:
+        """Calculate cloud cover for a given slice, group or layer.
```
Comment (loforest) [inline src/ampycloud/data.py:418]:
```suggestion
        """Calculate cloud amount for a given slice, group or layer.
```

File: src/ampycloud/data.py [hunk 9]
```diff
@@ -354,36 +391,47 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
...
+            original_ids (npt.ArrayLike): the original IDs of the slices/ groups/
+                layers.
 
+        Returns:
+            pd.DataFrame: The input data frame with results in the okta column.
+
+        Results are written to the "okta" column of the DF.
+
+        """
+        for ind, oid in enumerate(original_ids):
```
Comment (loforest) [inline src/ampycloud/data.py:432]: Can this function be used also to recalculate the amount from the merged groups/layers? (i.e. not the original_ids)

File: src/ampycloud/data.py [hunk 10]
```diff
@@ -395,35 +443,50 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
...
-                              base_lvl_alt_perc)
+        return pdf
+
+    def _add_sligrolay_information(
+            self,
+            which: str,
+            pdf: pd.DataFrame,
+            original_ids: npt.ArrayLike
+        ) -> pd.DataFrame:
+        """Add additional information for slices/ groups/ layers.
```
Comment (loforest) [inline src/ampycloud/data.py:469]:
```suggestion
        """Add statistical properties to slices/ groups/ layers .
```

File: src/ampycloud/data.py [hunk 11]
```diff
@@ -438,7 +501,92 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
                 boost=self.prms['GROUPING_PRMS']['fluffiness_boost'],
                 **self.prms['LOWESS'])
 
-            # Finally, create the METAR-like code for the cluster
+        return pdf
+
+    def _calculate_sligrolay_base_height(
```
Comment (loforest) [inline src/ampycloud/data.py:506]: As for the function `_calculate_cloud_cover`, how can this be done for the merged groups/layers?

File: src/ampycloud/data.py [hunk 12]
```diff
@@ -514,9 +662,65 @@ def find_slices(self) -> None:
...
+        self.metarize(
+            which='slices',
+        )
+
+    def _merge_close_groups(self) -> None:
+        """Merge groups that are closer than the minimum separation at their
+        respective altitudes.
+
+        Algorithm steps:
+            1. Calculate preliminery group base altitude
```
Comment (loforest) [inline src/ampycloud/data.py:674]:
```suggestion
            1. Calculate preliminary group base altitude
```

File: src/ampycloud/data.py [hunk 13]
```diff
@@ -514,9 +662,65 @@ def find_slices(self) -> None:
...
+        base_alt_diffs = prelim_groups['alt_base'].diff()
+        # create a boolean series to select values, fillna is necessary as
+        # first entry of diff will be nan
+        lt_min_sep_indexer = (base_alt_diffs < min_seps_grp).fillna(False)
+        while len(prelim_groups[lt_min_sep_indexer]) > 0:
+            # start with the two lowest layers that are too close
+            idx = prelim_groups[lt_min_sep_indexer].index[0]
+            # set the group id in the ceilo data to id of group below
+            data_idxer = self.data['group_id'] == prelim_groups['original_id'].loc[idx]
+            self.data.loc[data_idxer, 'group_id'] = prelim_groups['original_id'].iloc[idx - 1]
```
Comment (loforest) [inline src/ampycloud/data.py:709]: I was expecting this to be idx + 1 since idx is set to the loest layer at line 706.

File: src/ampycloud/data.py [hunk 14]
```diff
@@ -514,9 +662,65 @@ def find_slices(self) -> None:
...
+            # drop the group
+            prelim_groups.drop(index=idx, inplace=True)
+            # resetting because we must not have index gaps in the next iteration
+            prelim_groups.reset_index(drop=True, inplace=True)
+            # now we recalculate the base alt for the merged supergroup
```
Comment (loforest) [inline src/ampycloud/data.py:714]: Should we recalculate the amount too?

File: src/ampycloud/data.py [hunk 15]
```diff
@@ -514,9 +662,65 @@ def find_slices(self) -> None:
...
 
         # Finally, let's metarize these slices !
-        self.metarize(which='slices', base_lvl_alt_perc=self.prms['BASE_LVL_ALT_PERC'],
-                      base_lvl_lookback_perc=self.prms['BASE_LVL_LOOKBACK_PERC'],
-                      lim0=self.prms['MAX_HITS_OKTA0'], lim8=self.prms['MAX_HOLES_OKTA8'])
+        self.metarize(
+            which='slices',
+        )
+
+    def _merge_close_groups(self) -> None:
```
Comment (loforest) [inline src/ampycloud/data.py:669]: This function is not for the faint hearted. It is important to test it properly and ensure we treat possible expections and corner cases.
Also, since it is specific to groups, why don't we apply this to layers too?

**Review by loforest — COMMENTED** [review]
Finished my review!
There is one small change in the test_data.py (incorrect size of 'MIN_SEP_LIMS' array) and a possible issue in the function `_calc_base_alt` related to the lookback_perc in case the height values are not sorted according to time.
The remainder are only suggestions to improve docstrings and questions to understand what the code does and challenge potential issues.

**Comment by loforest** [comment]
> > > I wonder if we could configure a linter to catch that automatically. If you know @fpavogt let me know, otherwise I'll check tomorrow.
> > 
> > 
> > Sorry @regDaniel, I've got no clue on this.
> 
> Ok I found two things:
> 
> 1. We can enforce using type hints by passing `--disallow-untyped-defs` to mypy.
> 2. We do not use mypy in the actions workflow. Sorry, I was implicitly assuming we use it already and I would strongly encourage us to do so. If I run it locally I get quite some static type checking errors. Those are not necessarily bugs but we need to fix this, I'll address the most severe ones now and open an issue for the rest. @loforest if we fix typing errors at a later stage, can we deliver a patch release without CN?

1. These might be two other ways to do it:
   - Can we add the type hints check to the list of errors after --restrict E here? https://github.com/MeteoSwiss/ampycloud/blob/develop/.github/workflows/CI_pylinter.yml
   - In pysteps they use pre-commit hooks to block developers from pushing code that is not formatted properly (it might work also for checking type hints): https://github.com/pySTEPS/pysteps/blob/master/.pre-commit-config.yaml
2. We will likely do another ampycloud release during 2024, e.g. to improve the residual ceiling false alarms when the ceilometers are inside FG (if possible). We will include in the same release the type hints checks as Standard Change (Software Refactoring). It is too much overhead to do a Standard Change CN only for this, but if in case there are no new features for a while, we might consider to do just a software refactoring CN.

File: src/ampycloud/data.py (hunk 14 above)
Comment (regDaniel) [reply src/ampycloud/data.py:714]: I think, it is not necessary here, as we call the full `metarize` method after the merging step.

File: src/ampycloud/data.py (hunk 13 above)
Comment (regDaniel) [reply src/ampycloud/data.py:709]: yep, it's confusing I agree. The reason is the following. the `.diff()` method of pandas assigns the difference of two values in the series to the minuend, i.e. diff(k) = ser(k) - ser(k-1). When you then compare the calculated difference to the minimum separation, you'll get the true value at k. But you want to merge groups k-1 and k. Is that answer understandable? Sorry, it's a bit of a 🤯 and it's not impossible I got it wrong in the end.

File: test/ampycloud/test_data.py [hunk 16]
```diff
@@ -139,12 +139,42 @@ def test_ceilochunk_basic():
...
+    reset_prms()
+
 def test_bad_layer_sep_lims():
     """ Test that giving problematic layer separation limits does raise an error. """
 
     # Make sure that bad layering min sep values raises an error
-    dynamic.AMPYCLOUD_PRMS['LAYERING_PRMS']['min_sep_vals'] = [150, 1000]
-    dynamic.AMPYCLOUD_PRMS['LAYERING_PRMS']['min_sep_lims'] = [5000, 10000]
+    dynamic.AMPYCLOUD_PRMS['MIN_SEP_VALS'] = [150, 1000]
+    dynamic.AMPYCLOUD_PRMS['MIN_SEP_LIMS'] = [5000, 10000]
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply test/ampycloud/test_data.py:177]: it is raised as soon as we use the parameter when we want to get the minimum separation for a given height. I would actually prefer to do a refactoring with all parameter to enforce parameter types and properties with the use of pydantic and data classes that hold the parameter.

File: src/ampycloud/data.py [hunk 17]
```diff
@@ -257,67 +258,94 @@ def max_hits_per_layer(self) -> int:
...
+            float: The minimum separation for the given altitude.
+
+        Raises:
+            AmpycloudError: If the length of MIN_SEP_LIMS is not one less than
+                MIN_SEP_VALS
 
         """
+        if len(self.prms['MIN_SEP_LIMS']) != \
+            len(self.prms['MIN_SEP_VALS']) - 1:
+                raise AmpycloudError(
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply src/ampycloud/data.py:303]: see answer above. I think we should  use pydantic.

File: src/ampycloud/data.py (hunk 6 above)
Comment (regDaniel) [reply src/ampycloud/data.py:411]: I'm pretty sure yes. I wanted to avoid writing a second for loop so I put it here. But I could also put it into the loop before the `if`. the way I see it, the two blocks do however not depend on each other.

File: src/ampycloud/data.py (hunk 9 above)
Comment (regDaniel) [reply src/ampycloud/data.py:432]: not sure I get the comment. This function is used in the `metarize()` public method to calculate the cloud amount and `metarize()` is called after the merging steps.

File: src/ampycloud/data.py (hunk 11 above)
Comment (regDaniel) [reply src/ampycloud/data.py:506]: It can be done at any time if we have a group/ layer data frame and a data frame holding the original IDs. The non-correspondence between original IDs (the IDs in the data) and the indices in the functions are a source of confusion. But if we want to change this logic, I'd have to think quite a bit, how we could achieve it as we rely on the sorting and reindexing that causes this non-correspondence.

**Comment by regDaniel** [comment]
> > > > I wonder if we could configure a linter to catch that automatically. If you know @fpavogt let me know, otherwise I'll check tomorrow.
> > > 
> > > 
> > > Sorry @regDaniel, I've got no clue on this.
> > 
> > 
> > Ok I found two things:
> > 
> > 1. We can enforce using type hints by passing `--disallow-untyped-defs` to mypy.
> > 2. We do not use mypy in the actions workflow. Sorry, I was implicitly assuming we use it already and I would strongly encourage us to do so. If I run it locally I get quite some static type checking errors. Those are not necessarily bugs but we need to fix this, I'll address the most severe ones now and open an issue for the rest. @loforest if we fix typing errors at a later stage, can we deliver a patch release without CN?
> 
> 1. These might be two other ways to do it:
>    
>    * Can we add the type hints check to the list of errors after --restrict E here? https://github.com/MeteoSwiss/ampycloud/blob/develop/.github/workflows/CI_pylinter.yml
>    * In pysteps they use pre-commit hooks to block developers from pushing code that is not formatted properly (it might work also for checking type hints): https://github.com/pySTEPS/pysteps/blob/master/.pre-commit-config.yaml
> 2. We will likely do another ampycloud release during 2024, e.g. to improve the residual ceiling false alarms when the ceilometers are inside FG (if possible). We will include in the same release the type hints checks as Standard Change (Software Refactoring). It is too much overhead to do a Standard Change CN only for this, but if in case there are no new features for a while, we might consider to do just a software refactoring CN.

Thanks for your answers.
1.1. yep that would be an option. But we can also invoke mypy directly in the actions workflow. Personally I am always in favor of using as little wrapper scripts as possible. The standard tools (mypy, pylint) are anyways invoked in good IDEs if you have them in your dev environment.
1.2. I'm not in favor of pre-commit hooks as you then have to make sure that they behave exactly the same as the linting done in you pipeline and in my experience, this is a non-trivial task which needs quite a bit of regular maintenance. For me it's ok to do these checks on the level of the CI/CD pipeline exclusively. If devs do not use a proper IDE/ do not run the linters locally, they'll have to do more fixing once they try to open merge requests.
2. Ok thanks, I guess it could go in a separate release together with some tuning maybe?

File: src/ampycloud/data.py (hunk 9 above)
Comment (regDaniel) [reply src/ampycloud/data.py:432]: maybe rename `original_id` to `cluster_id`

Comment (fpavogt) [reply src/ampycloud/data.py:432]: I would support a renaming. The term "original ID" is definitely not very explicit. ``cluster_id`` sounds good.

File: test/ampycloud/test_data.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (regDaniel) [reply test/ampycloud/test_data.py:177]: here, we need to leave it as is as this is a test if the raise works properly.

File: src/ampycloud/layer.py [hunk 18]
```diff
@@ -138,10 +139,46 @@ def best_gmm(abics: np.ndarray, mode: str = 'delta',
...
+        lookback_perc (int): Percentage of points to take into account. 100% would
+            correspond to all points, 50% to the recent half, etc.
+        alt_perc (int): Percentage of points that should be neglected when calculating
+            the base height. Base height will be the minimum of the remaining points.
+
+    Returns:
+        float: The layer base altitude.
+
+    Raises:
+        AmpycloudError: If the array passed to the n_largest percentile calculation
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply src/ampycloud/layer.py:161]: rephrased sentence

File: src/ampycloud/layer.py [hunk 19]
```diff
@@ -240,8 +280,14 @@ def ncomp_from_gmm(vals: np.ndarray,
         return best_ncomp, best_ids, abics
 
     # If I found more than one component, let's make sure that they are sufficiently far apart.
-    # First, let's compute the mean component heights
-    mean_comp_heights = [np.mean(vals_orig[best_ids == i]) for i in range(ncomp[best_model_ind])]
+    # First, let's compute the component base altitude
+    mean_comp_heights = [
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply src/ampycloud/layer.py]: done

File: src/ampycloud/layer.py [hunk 20]
```diff
@@ -138,10 +139,46 @@ def best_gmm(abics: np.ndarray, mode: str = 'delta',
...
+    Returns:
+        float: The layer base altitude.
+
+    Raises:
+        AmpycloudError: If the array passed to the n_largest percentile calculation
+            is empty.
+
+    """
+    n_largest_idxs = vals.argsort()[
+        - int(len(vals) * lookback_perc / 100):
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply src/ampycloud/layer.py]: done

File: src/ampycloud/data.py (hunk 15 above)
Comment (regDaniel) [reply src/ampycloud/data.py:667]: added comment

File: src/ampycloud/data.py [hunk 21]
```diff
@@ -257,67 +258,94 @@ def max_hits_per_layer(self) -> int:
...
 
-        The slice/group/layer parameters computed/derived by this method include:
+        """
+        # Start computing the base altitude
+        # First, compute which points should be considered in terms of lookback time
+        n_to_use = int(np.floor(len(
+            self.data.loc[data_indexer]) * self.prms['BASE_LVL_LOOKBACK_PERC']/100
+        ))
+        # Then, actually compute the base altitude, possibly ignoring the lowest points
+        return np.percentile(
```
[the remark this answers is not in the document]
Comment (regDaniel) [reply src/ampycloud/data.py]: done with the refactoring

File: src/ampycloud/data.py (hunk 7 above)
Comment (regDaniel) [reply src/ampycloud/data.py:413]: done manually

File: src/ampycloud/data.py (hunk 8 above)
Comment (regDaniel) [reply src/ampycloud/data.py:416]: done manually

File: src/ampycloud/data.py (hunk 9 above)
Comment (regDaniel) [reply src/ampycloud/data.py]: renamed to `cluster_id`
